torch/distributed/_tensor/experimental/distribute_function.py

```diff
@@ -0,0 +1,90 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates
+
+import collections
+import importlib
+import types
+from typing import Callable, Dict, List, Optional, Set, Tuple
+
+import torch
+from torch import nn
+from torch.distributed.device_mesh import DeviceMesh
+
+
+_original_functions: Dict[Callable, Callable] = {}
+_wrapper_functions: Dict[Callable, Callable] = {}
+_replaced_objs: collections.defaultdict[
+    Callable, Set[Tuple[types.ModuleType, str]]
+] = collections.defaultdict(set)
+
+
+def distribute_function(
+    fn: Callable,
+    fn_module: types.ModuleType,
+    fn_callers: List[nn.Module],
```

Comment:
For these args, I am assuming that our main motivation is to try to patch the F.spda with this `distribute_function` so that there's no model code changes required, but I'm thinking that these args might be too much for user to use and understand, maybe a better way would just be: if user want to use CP, they can replace the F.spda with our annotated/distributed `ring_spda` function

Reply:
I agree this API is a complicated. We don't have to expose this API to users. I just use this API to implement the `enable_context_parallel()` in the next PR. `enable_context_parallel()` takes 3 argument, seq_dim, callers, mesh. `seq_dim` and `mesh` are easy to understand. `callers` is what nn.Module calls SDPA.